[WIP] feat: call python methods from forum v2

Faraz32123 opened the pull request:
**This PR is still a work in progress and not yet ready for final review.**

- Added new CourseWaffle flag named `ENABLE_FORUM_V2`. If it's not set for the course, V1(cs_comment_service) will be called.
- Now we are using our forum v2 python native APIs based on CourseWaffle, `ENABLE_FORUM_V2`.
- Modified old/current unit tests of edx-platform that uses forum APIs as there's now implementation of WaffleFla. So, had to modify the unit tests.
- Added forum to the edx-platform requirements.

Commit:
feat: call python methods from forum v2
- directly call python native APIs from forum v2 for pin, unpin thread,
commentables count_stats and get user's data by user_id
- add forum to the edx-platform requirements

regisb commented on openedx/core/djangoapps/django_comment_common/comment_client/course.py:
```diff
@@ -29,17 +30,8 @@ def get_course_commentable_counts(course_key: CourseKey) -> Dict[str, Dict[str,
-        metric_action='commentable_stats.retrieve',
-    )
-    return response
+    commentable_stats = forum_api.retrieve_commentables_stats(str(course_key))
```
We don't see the "retrieve" verb very often in APIs. Let's use "get" instead, i.e: "get_commentable_stats".

regisb commented on openedx/core/djangoapps/django_comment_common/comment_client/thread.py:
```diff
@@ -193,27 +194,11 @@ def unFlagAbuse(self, user, voteable, removeAll):
-        )
-        self._update_from_response(response)
+        thread_data = forum_api.pin_thread(user.id, thread_id)
+        self._update_from_response(thread_data)
```
I wonder how we can get rid of these obnoxious calls to `_update_from_response`. That method blindly updates all attributes:

    def _update_from_response(self, response_data):
        for k, v in response_data.items():
            if k in self.accessible_fields:
                setattr(self, k, v)
            else:
                log.warning(
                    "Unexpected field {field_name} in model {model_name}".format(
                        field_name=k,
                        model_name=self.__class__.__name__
                    )
                )

If the API call accidentally returns a field that is in "accessible_fields", then that field will be updated, and there is no way that unit tests will catch that error.

Faraz32123 replied:
I think, if the API returns a value that is in self.accessible_fields, it should set it as it was before. As accessible_fields are already defined for each model, it will only set those or otherwise log.
May be I don't get the point.

Commit:
feat: migrate some APIs to native python
- directly call python native APIs from forum v2 for get parent comment,
create parent comment and create child comment.
- rename retrieve_commentables_stats method to get_commentables_stats and
retrieve_user to get_user.

Commit:
feat: pass params to python native APIs
- refactored code and now pass proper parameters to python native APIs
instead of a single dict

Commit:
feat: code refactor and migrate delete comment API

ormsbee commented:
As you folks are building this out, please use a `CourseWaffleFlag` to switch between running the old code that hits the service from edx-platform and the new code that uses the `forum` app. When a migration this big rolls out on a site that runs off of master (like 2U or one of MIT's sites), it will usually go:

1. Turn it on for a few select courses to try it out, see if there are any major regressions.
2. Turn it on by default for everyone.
3. Turn it back off for a few select courses where weird edge case bugs arise.

We can turn this flag on by default for Sumac.

Commit:
fix: user tests

Commit:
feat: fix tests for get_user API
- get_user API tests are now passing in test_views.py and test_serializers.py
- add get_user api patch in all tests
- fix httppretty request count in some tests
- fix test_patch_read_non_owner_user test

Commit:
feat: use new coursewaffle flag to run old code
- add `ENABLE_FORUM_V2` course waffle flag to switch between old code i.e.
cs_comment_service and new code i.e. forum v2.
- mock course waffle flag is_enabled method i.e. ENABLE_FORUM_V2.is_enabled(),
so that old unit tests can be run and passed.
- refactor code(that parts of code whose native APIs are implemented till now)
where we call the native APIs

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Commit:
fix: resolve inconsistant python dependencies
- run `make compile-requirements forum`. Error: It appears that the
Python dependencies in this PR are inconsistent: A re-run of
`make compile-requirements` produced changes.
- fix quality checks

Commit:
fix: fix new failing tests
- fix new tests related to discussion that failed after fixing previous tests
these are failing due to no.of argument difference
https://github.com/openedx/edx-platform/actions/runs/11069160532/job/30756121710?pr=35490

Commit:
feat: migrate user active_thread api

Commit:
feat: migrate fetch subscription

Commit:
feat: add python native APIs from forum v2
- migrate following native APIs i.e. mark_thread_as_read, create_subscription,
delete_subscription, update_thread_votes, update_comment_votes, delete_thread_vote,
delete_comment_vote, get_user_threads, retire_user, update_username,
get_user_course_stats, update_users_in_course, flag/unflag comment/threads
from forum v2 to edx-platform
- refactor some code

Commit:
feat: update thread apis

Commit:
feat: add threads api

Commit:
fix: quality checks

Commit:
feat: migrate update_user api

taimoor-ahmed-1 commented on openedx/core/djangoapps/django_comment_common/comment_client/course.py:
```diff
@@ -29,17 +31,20 @@ def get_course_commentable_counts(course_key: CourseKey) -> Dict[str, Dict[str,
-        metric_action='commentable_stats.retrieve',
-    )
-    return response
+    if is_forum_v2_enabled(course_key):
+        commentable_stats = forum_api.get_commentables_stats(str(course_key))
+    else:
+        url = f"{settings.PREFIX}/commentables/{course_key}/counts"
+        commentable_stats = perform_request(
+            'get',
+            url,
+            metric_tags=[
+                f"course_key:{course_key}",
+                "function:get_course_commentable_counts",
+            ],
+            metric_action='commentable_stats.retrieve',
+        )
+    return commentable_stats
```
the if-else condition seems quite redundant, maybe we can use a method decorator to achieve the same task. This will reduce git difference and make the mode easier to read.
```
def forum_v2_decorator(forum_v2_func):
    def decorator(func):
        def wrapper(course_key: CourseKey, *args, **kwargs):
            if is_forum_v2_enabled(course_key):
                return forum_v2_func(course_key, *args, **kwargs)
            else:
                return func(course_key, *args, **kwargs)
        return wrapper
    return decorator

@forum_v2_decorator(forum_api.get_commentables_stats)
def get_course_commentable_counts(course_key: CourseKey) -> Dict[str, Dict[str, int]]:
    url = f"{settings.PREFIX}/commentables/{course_key}/counts"
    commentable_stats = perform_request(
        'get',
        url,
        metric_tags=[
            f"course_key:{course_key}",
            "function:get_course_commentable_counts",
        ],
        metric_action='commentable_stats.retrieve',
    )
    return commentable_stats

```

Faraz32123 replied:
I will be avoiding this change for now, as the decorator method will become complex for handling native requests for all APIs. 
Note: As we will remove the else condition to V1 in future

Commit:
feat: retrieve course_id

Commit:
fix: course_id issue for course_waffle flag

Commit:
fix: temp

Commit:
chore: code refactor, fix quality checks

Commit:
feat: add tests for native APIs
in this commit:
- migrated the tests for native APIs from test_serializers to test_serializers_native_views file.
- tests from test_views to test_views_native_views is still in progress
- fixed edx-platform tests which were failing after we added get course_id APIs
- fixed course_key error for user's retrieve APIs

Commit:
feat: add group_id actions tests

Commit:
feat: new native APIs tests progress

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Commit:
chore: move native API tests to separate PR

Commit:
fix: CI checks

Commit:
chore: change the forum branch to master

Commit:
fix: older edx-platform tests that uses V1
- Before resolving course_id availability issue for all APIs to be used in
CourseWaffle flag, only 4 files tests were failing, We fixed those.
Now due to latest fixes, some more tests were failing that uses v1, this PR will
fix those issues.

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Commit:
feat: pass course_id to native APIs
course_id is needed to be used in second coursewaffle flag

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Commit:
fix: update functions to named arguments

Commit:
fix: pylint issues

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Commit:
fix: tests, add mocks for getting course_id APIs

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Commit:
fix: failing unit tests

regisb commented on openedx/core/djangoapps/django_comment_common/comment_client/user.py:
```diff
@@ -144,57 +229,75 @@ def _retrieve(self, *args, **kwargs):
+                response = forum_api.get_user(self.attributes["id"], retrieve_params)
+            except ForumV2RequestError as e:
+                self.save({"course_key": course_key})
+                response = forum_api.get_user(self.attributes["id"], retrieve_params)
```
This was causing an error on my computer, where `self.attributes["id"]` was None. Strangely, the error has resolved itself.

What is this strange pattern of catching an exception to run the same `get_user` function one more time?

Faraz32123 replied:
actually we followed the same pattern how edx-platform was calling the V1 APIs before.
At first, it tries to get the user, if it doesn't exists. It then creates the user in forum APP and then again gets.
For the new user that goes to discussion MFE, this error always occur once.
You can also see the else part for V1 behaviour.

regisb replied:
Got it! Sorry!

regisb commented on openedx/core/djangoapps/django_comment_common/comment_client/models.py:
```diff
@@ -208,3 +215,169 @@ def url(cls, action, params=None):
+            body = request_data["body"]
+            user_id = str(request_data["user_id"])
+        except KeyError as e:
+            raise e
```
This is raising an issue on my computer:

```
...
lms-1  |   File "/openedx/edx-platform/lms/djangoapps/discussion/rest_api/views.py", line 667, in create
lms-1  |     return Response(create_thread(request, request.data))
lms-1  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lms-1  |   File "/openedx/edx-platform/lms/djangoapps/discussion/rest_api/api.py", line 1482, in create_thread
lms-1  |     serializer.save()
lms-1  |   File "/openedx/venv/lib/python3.11/site-packages/rest_framework/serializers.py", line 212, in save
lms-1  |     self.instance = self.create(validated_data)
lms-1  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
lms-1  |   File "/openedx/edx-platform/lms/djangoapps/discussion/rest_api/serializers.py", line 467, in create
lms-1  |     thread.save()
lms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/django_comment_common/comment_client/models.py", line 169, in save
lms-1  |     response = self.handle_create(params)
lms-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^
lms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/django_comment_common/comment_client/models.py", line 325, in handle_create
lms-1  |     response = self.handle_create_thread(str(course_key))
lms-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/django_comment_common/comment_client/models.py", line 367, in handle_create_thread
lms-1  |     raise e
lms-1  |   File "/openedx/edx-platform/openedx/core/djangoapps/django_comment_common/comment_client/models.py", line 365, in handle_create_thread
lms-1  |     user_id = str(request_data["user_id"])
lms-1  |                   ~~~~~~~~~~~~^^^^^^^^^^^
lms-1  | KeyError: 'user_id'
```

Why are we catching the KeyError and raising it again?

Faraz32123 replied:
yeah point, we can just remove the try catch.
Actually, might be forum master branch is not that stable yet. 
We tested the flow with forum v2(on these branches: feat/add_tests_for_forum_v2_native_python_calls and then on alisalman/refactoring).
But this'll also fail now as we have changed the signatures for native APIs.

Faraz32123 replied:
Also the backend is set to mysql in forum master branch.
Plus let me see this user_id issue, this shouldn't occur as it is a required field.
if u switch to this commit([5aa42da](https://github.com/openedx/edx-platform/pull/35490/commits/5aa42dafbcfe0d9d6ee87171e7edf525f92e2ea6)) in this PR. This User Id does exists.

Faraz32123 replied:
actually i did fix this error [here](https://github.com/edly-io/forum/pull/100/commits/c8af1504e3332665712175b755e09c8667856655). 
made a new PR [here](https://github.com/edly-io/forum/pull/105).
User_id issue will be resolved after merge.

Commit:
refactor: remove unnecessary try catch

regisb commented:
The goal of this PR is to integrate the [new forum application](https://github.com/edly-io/forum/) into edx-platform. In practice, this means the following changes:

- [x] Introduce a course waffle flag "forum_v2.enable_forum_v2".
- [x] Courses for which this waffle flag is enabled will make use of the new forum app. In practice, it's possible to have both the legacy forum v1 and the new v2 app coexist with one another, on a course per course basis.
- [x] The forum v2 app will default to storing data in mongodb. But to use mysql as a data storage (the default in tutor), you should use the "forum_v2.enable_mysql_backend" course waffle flag.
- [x] To migrate data on a course-per-course basis from mongodb to mysql, you should use the "forum_migrate_course_from_mongodb_to_mysql" management command.

Note that this PR does not include all unit tests for the forum v2 native API. This is because migrating unit tests is taking much longer than expected. We will take out this PR from draft as soon as it is considered production-ready, despite the fact that some code in edx-platform might not be 100% covered.

regisb commented on requirements/edx/base.txt:
```diff
@@ -227,6 +227,7 @@ django==4.2.16
     #   enmerkar
     #   enmerkar-underscore
     #   event-tracking
+    #   forum
```
I think you forgot to commit the changes in base.in. This might be because of a conflict resolution.

Faraz32123 replied:
Actually there are no changes in base.in, Forum requirement is set in github.in

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

Faraz32123 committed:
Merge branch 'master' into feat/migrate_APIs_from_http_to_python_call

regisb commented:
My branch rename as caused this PR to close... Sorry about that. Can you please open another PR @Faraz32123? Likewise for https://github.com/edly-io/edx-platform/pull/585

Faraz32123 commented:
> My branch rename as caused this PR to close... Sorry about that. Can you please open another PR @Faraz32123? Likewise for [edly-io#585](https://github.com/edly-io/edx-platform/pull/585)

Yeah, i noticed that. I'll sure create a new PR.

regisb commented:
New PR: https://github.com/openedx/edx-platform/pull/35671
